File: v1/src/main/java/com/api/v1/car/utils/CarFinderUtilImpl.java
package com.api.v1.car.utils;


import com.api.v1.car.annotations.VIN;
import com.api.v1.car.domain.Car;
import com.api.v1.car.domain.CarRepository;
import com.api.v1.car.exceptions.CarNotFoundException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

@Component
public class CarFinderUtilImpl implements CarFinderUtil {

    @Autowired
    private CarRepository repository;

    @Override
    public Mono<Car> find(@VIN String vin) {
        return repository
                .findAll()
                .filter(e -> e.getVin().equals(vin))
                .singleOrEmpty()
                .switchIfEmpty(Mono.error(new CarNotFoundException(vin)));
    }

}

File: v2/src/main/kotlin/com/api/v2/car/CarFinderUtilImpl.kt
package com.api.v2.car

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
private class CarFinderUtilImpl: CarFinderUtil {

    @Autowired
    private lateinit var carRepository: CarRepository

    override suspend fun find(vin: String): Car {
        return withContext(Dispatchers.IO) {
            val existingCar = carRepository
                .findAll()
                .filter { e -> e.vin ==  vin }
                .firstOrNull()
            if (existingCar == null) {
                throw CarNotFoundException(vin)
            }
            existingCar
        }

    }

}